spring-cloud-data-module-deployers/spring-cloud-data-module-deployer-cloudfoundry/src/main/java/org/springframework/cloud/data/module/deployer/cloudfoundry/AbstractPaginatedResponse.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.data.module.deployer.cloudfoundry;
+
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+abstract class AbstractPaginatedResponse {
+
+	private volatile Integer totalResults;
```

Comment:
> I was encouraged to use volatile when other pieces of hidden code update/read the field.

Volatile is not meant to be used when hidden code (such as serialization) is mutating fields - it is meant to be used to enforce _visibility_ of mutations across threads.

> In general it does no harm to assume that for any field.

In practice there is low overhead for mutating volatile fields - the cost isn't zero but it's pretty low:

http://stackoverflow.com/questions/4633866/is-volatile-expensive

The issue for me is that when I see `volatile` it is sending a message to me that the class was **explicitly designed** to be thread safe. 

For instance, if we take a look at `ApplicationInstanceStatus.Statistics`:

```
private volatile Usage usage;
private volatile String name;
private volatile List<String> uris;
private volatile String host;
private volatile Integer port;
private volatile Integer uptime;
private volatile Integer memoryQuota;
private volatile Integer diskQuota;
private volatile Integer fdsQuota;
```

Do we have a use case in mind where a thread will be updating a field (let's say `port`) while another thread updates another field (let's say `uptime`)? Would an instance of this class be considered "usable" if its fields are still being populated? Or is it expected that once all of the fields are populated, _that_ is when this object should be used?

I suspect it is the latter - which means having all of the fields `volatile` isn't useful if we expect that _all_ of the fields will be populated before the object is consumed (i.e. serialized as part of a request, rendered on a web page, etc).

I expect usage of instances of this class to be as follows (please correct me if I'm wrong):

1. Request is made to the admin server to obtain statistics
2. Admin server makes a request to CF to obtain statistics on the same thread that received the initial request (same thread as 1)
3. Admin server creates a new instance of `Statistics` and populates it (same thread as 1)
4. Admin server serializes the instance and sends it as part of the response (same thread as 1)

Given this usage pattern, the object will never be exposed to multiple threads.

If for some reason in the future it were exposed to multiple threads, it would be up to those threads to properly synchronize their usage of this object to make sure that one thread isn't performing a partial read (i.e. only some of the fields are populated) while another thread is updating the fields. Having the fields `volatile` doesn't help with that.

> In my opinion this oughtn't to be necessary for spring/jackson accessed fields as there ought to be correct synchronisation observed in those implementations. 

Exactly!!

> But just in case, Ben Hale got me to put those in there.

Here is where I disagree - IMO we should avoid doing things "just in case" and instead we should be _explicit_ about expected usage of our code.